frontend/src/pages/CompareV2.tsx

```diff
@@ -14,13 +14,166 @@
  * limitations under the License.
  */
 
-import React from 'react';
-import { commonCss } from 'src/Css';
+import React, { useEffect, useRef, useState } from 'react';
+import { useQuery } from 'react-query';
+import { ApiRunDetail } from 'src/apis/run';
+import Hr from 'src/atoms/Hr';
+import Separator from 'src/atoms/Separator';
+import CollapseButton from 'src/components/CollapseButton';
+import { QUERY_PARAMS, RoutePage } from 'src/components/Router';
+import { commonCss, padding } from 'src/Css';
+import { Apis } from 'src/lib/Apis';
+import Buttons from 'src/lib/Buttons';
+import { URLParser } from 'src/lib/URLParser';
+import { errorToMessage } from 'src/lib/Utils';
+import { classes, stylesheet } from 'typestyle';
+import { PageProps } from './Page';
+import RunList from './RunList';
+
+const css = stylesheet({
+  outputsRow: {
+    marginLeft: 15,
+    overflowX: 'auto',
+  },
+});
+
+export const overviewSectionName = 'Run overview';
+export const paramsSectionName = 'Parameters';
+export const metricsSectionName = 'Metrics';
+
+function CompareV2(props: PageProps) {
+  const { updateBanner, updateToolbar } = props;
+
+  const runlistRef = useRef<RunList>(null);
+  const [collapseSections, setCollapseSections] = useState<{ [key: string]: boolean }>({});
+  const [selectedIds, setSelectedIds] = useState<string[]>([]);
+
+  const queryParamRunIds = new URLParser(props).get(QUERY_PARAMS.runlist);
+  const runIds = (queryParamRunIds && queryParamRunIds.split(',')) || [];
+
+  // Retrieves run details.
+  const { isError, data, refetch } = useQuery<ApiRunDetail[], Error>(
+    ['run_details', { ids: runIds }],
+    () => Promise.all(runIds.map(async id => await Apis.runServiceApi.getRun(id))),
+    {
+      staleTime: Infinity,
+      onError: async error => {
+        const errorMessage = await errorToMessage(error);
+        updateBanner({
+          additionalInfo: errorMessage ? errorMessage : undefined,
+          message: `Error: failed loading ${runIds.length} runs. Click Details for more information.`,
+          mode: 'error',
+        });
+      },
+      onSuccess: () => props.updateBanner({}),
+    },
+  );
+
+  useEffect(() => {
+    const refresh = async () => {
+      if (runlistRef.current) {
+        await runlistRef.current.refresh();
+      }
+      await refetch();
+    };
+
+    const buttons = new Buttons(props, refresh);
+    updateToolbar({
+      actions: buttons
+        .expandSections(() => setCollapseSections({}))
+        .collapseSections(() =>
+          setCollapseSections({
+            [overviewSectionName]: true,
+            [paramsSectionName]: true,
+            [metricsSectionName]: true,
+          }),
+        )
+        .refresh(refresh)
+        .getToolbarActionMap(),
+      breadcrumbs: [{ displayName: 'Experiments', href: RoutePage.EXPERIMENTS }],
+      pageTitle: 'Compare runs',
+    });
+  }, [updateToolbar, refetch]);
```

Comment:
I did not include `props` in the `useEffect` dependency list as that would cause an infinite re-rendering of the page, as `updateToolbar` is called. This is also the approach used on the `NewRunV2.tsx` page: https://github.com/kubeflow/pipelines/blob/468d78046528f35b623b9c2eb3cc10fe5558e657/frontend/src/pages/NewRunV2.tsx#L125-L131

However, this still results in an `exhaustive-deps` warning:
![ExhaustiveDepsWarning](https://user-images.githubusercontent.com/7987279/173158695-a79ac524-631b-41b4-85d8-e7f4be1de9e1.png)

Is there a solution I should use here to remove this warning, or should I simply leave it as-is?

Reply:
I guess you probably need to run this piece of `useEffect` logic once during mounting. Correct? If that is the case, you can remove the whole dependency list from `useEffect`: https://reactjs.org/docs/hooks-effect.html

```
  useEffect(() => {
    const refresh = async () => {
      if (runlistRef.current) {
        await runlistRef.current.refresh();
      }
      await refetch();
    };

    const buttons = new Buttons(props, refresh);
    updateToolbar({
      actions: buttons
        .expandSections(() => setCollapseSections({}))
        .collapseSections(() =>
          setCollapseSections({
            [overviewSectionName]: true,
            [paramsSectionName]: true,
            [metricsSectionName]: true,
          }),
        )
        .refresh(refresh)
        .getToolbarActionMap(),
      breadcrumbs: [{ displayName: 'Experiments', href: RoutePage.EXPERIMENTS }],
      pageTitle: 'Compare runs',
    });
  });
```

Reply:
If I don't include any `useEffect` dependencies, the page undergoes an infinite refresh with this error:
![useEffect Maximum Update Depth](https://user-images.githubusercontent.com/7987279/173422894-4282b8db-64fc-4b04-b0c2-edcbda7b0899.png)

I think it's because I have the `setCollapseSections` call within the `useEffect`, and that error mentions how a dependency array is necessary in those cases. I do think I just need to run the `useEffect` once upon mounting, but the dependency array seems required despite this?

If I then include the dependency array, what should I do about `props` and the warning that arises from leaving that out (even though the code works as expected, it generates a warning from `npm start`).

Reply:
Make sense, can you try to include an empty array `[]` to the `useEffect` dependency? Reference: https://stackoverflow.com/a/54923969

Reply:
That code works (same as the initial code), but results in this more comprehensive `exhaustive-deps` warning:

![ReactExhaustiveDeps](https://user-images.githubusercontent.com/7987279/173679613-867bac28-d62b-4563-9cba-9dc3ac94d60d.png)

I believe whenever some dependency is added to the `useEffect`, React will check on compilation if those dependencies are exhaustive.

Reply:
At the current point, I will advice to add this comment to `useEffect` until we get a better solution to avoid infinite re-rendering:

```
  // eslint-disable-next-line react-hooks/exhaustive-deps
```

Reference: https://bobbyhadz.com/blog/react-hook-useeffect-has-missing-dependency

Reply:
Sounds good - done! ✅  https://github.com/kubeflow/pipelines/pull/7882/commits/01a4bd5bc411ae09c1e785a48cd0d14231965962

Reply:
Since we still want to run this piece of logic only once, can you include an empty array `[]` to the `useEffect` dependency?

Reply:
Yes - done here: https://github.com/kubeflow/pipelines/pull/7882/commits/73494d79a08971c2c8fd8101817a647fb53b5be2. Thanks!